README.md
qqplotr
================
<a href="https://travis-ci.org/aloy/qqplotr"> <img border="0" src="https://travis-ci.org/aloy/qqplotr.svg?branch=master"> </a> <a href="https://cran.r-project.org/package=qqplotr"> <img border="0" src="https://www.r-pkg.org/badges/version/qqplotr"> </a>

-   [Installation](#installation)
-   [Details](#details)
    -   [Q-Q plot](#q-q-plot)
    -   [P-P plot](#p-p-plot)
-   [Usage](#usage)
    -   [Q-Q plot](#q-q-plot-1)
    -   [P-P plot](#p-p-plot-1)
-   [Shiny App](#shiny-app)
-   [References](#references)

<!-- README.md is generated from README.Rmd. Please edit that file -->
The `qqplotr` package extends some `ggplot2` functionalities by permitting the drawing of both quantile-quantile (Q-Q) and probability-probability (P-P) points, lines, and confidence bands. The functions of this package also allow a detrend adjustment of the plots, proposed by Thode (2002) to help reduce visual bias when assessing the results.

Installation
------------

If you would like to install the development version of `qqplotr`, you may do so by using, for example, `devtools`:

``` r
# install.packages("devtools")
library(devtools)
devtools::install_github("aloy/qqplotr")
```

If, instead, you wish to install the stable CRAN version, then simply do:

``` r
install.packages("qqplotr")
```

Details
-------

The functions of this package, implemeneted as Stats from `ggplot2`, are divided into two groups: (1) Q-Q and (2) P-P plot.

Both groups are composed of three functions: *point*, *line*, and *band*. Those Stats complement each other when drawn together, but they may also be plotted independently.

Below we will give an overview of all those Stats and, further in the document, we will present some usage examples.

### Q-Q plot

-   `stat_qq_point` This is a modified version of `ggplot2::stat_qq` with some parameters adjustments and a new option to detrend the points.
-   `stat_qq_line` Draws a reference line based on the data quantiles, as in `stats::qqline`.
-   `stat_qq_band` Draws confidence bands based on three methods: `"normal"`, `"boot"`, and `"ts"`:
    -   `"normal"` constructs simultaneous confidence bands based on Normal confidence intervals;
    -   `"boot"` creates pointwise confidence bands based on a parametric boostrap;
    -   `"ts"` constructs tail-sensitive confidence bands, as proposed by Aldor-Noiman et al. (2013).

In order to facilitate the visualization of multiple Q-Q band methods at the same time, the `geom_qq_band` Geom was also implemented. Its usage will be illustrated further below.

### P-P plot

-   `stat_pp_point` Plots cumulative probabilities versus probability points. The cumulative probability function is constructed with the sample data, and then evaluated at each probability point.
-   `stat_pp_line` Draws a reference identity line (*x* = *y*).
-   `stat_pp_band` Draws confidence bands. For now, only the bootstrap version (`"boot"`) is available.

Usage
-----

### Q-Q plot

Start by loading the `qqplotr` package:

``` r
require(qqplotr)
```

Let's start by simulating from a standard Normal distribution:

``` r
set.seed(0)
smp <- data.frame(norm = rnorm(100))
```

Then, we use the provided `stat_qq_*` functions to construct a complete Q-Q plot with the points, reference line, and the confidence bands. As default, the standard Q-Q Normal plot with Normal confidence bands is constructed:

``` r
gg <- ggplot(data = smp, mapping = aes(sample = norm)) +
    stat_qq_band() +
    stat_qq_line() +
    stat_qq_point() +
    labs(x = "Theoretical Quantiles", y = "Sample Quantiles")
gg
```

<img src="man/figures/README-unnamed-chunk-6-1.png" style="display: block; margin: auto;" />

As we can see, all the points lie within the confidence bands, which is expected for the given distribution.

As previously described in the Details section, three confidence bands constructs are available, which may be adjusted with the `bandType` parameter. Here, we may use the `geom_qq_band` instead of `stat_qq_band`, which permits a little more flexibility with the graphical parameters when constructing and visualizing different confidence bands.

``` r
gg <- ggplot(data = smp, mapping = aes(sample = norm)) +
    geom_qq_band(bandType = "ts", mapping = aes(fill = "TS")) +
    geom_qq_band(bandType = "normal", mapping = aes(fill = "Normal")) +
    geom_qq_band(bandType = "boot", mapping = aes(fill = "Bootstrap")) +
    stat_qq_line() +
    stat_qq_point() +
    labs(x = "Theoretical Quantiles", y = "Sample Quantiles") +
    scale_fill_discrete("Bandtype")
gg
```

<img src="man/figures/README-unnamed-chunk-7-1.png" style="display: block; margin: auto;" />

To construct Q-Q plots with other theoretical distributions we may use the `distribution` parameter. Specific distributional parameters may be passed as a list to the `dparams` paramater.

> Note that distributional parameters have little impact when building Q-Q plots, as changing them will only modify the x-axis range. In contrast, those paramaters will have a higher effect on P-P plots.

Now, let's use draw the Q-Q plot functions for the *mean ozone levels* from the `airquality` dataset . Since the data is non-negative, lets choose the Exponential distribution (`exp`) as the theoretical.

> It is important to note that the distribution nomenclature follows that from the `stats` package. So, if you wish to provide a custom distribution, you may do so by creating the density, cumulative, quantile, and random functions following the standard nomenclature from the `stats` package, i.e., for the `"custom"` distribution, you must define `"dcustom"`, `"pcustom"`, `"qcustom"`, and `"rcustom"` functions.

That being said, let's set `distribution = "exp"` and `rate = 2` (the latter one just to exemplify the usage of `dparams`):

``` r
di <- "exp" # exponential distribution
dp <- list(rate = 2) # exponential rate parameter

gg <- ggplot(data = airquality, mapping = aes(sample = Ozone)) +
    stat_qq_band(distribution = di, dparams = dp) +
    stat_qq_line(distribution = di, dparams = dp) +
    stat_qq_point(distribution = di, dparams = dp) +
    labs(x = "Theoretical Quantiles", y = "Sample Quantiles")
gg
```

<img src="man/figures/README-unnamed-chunk-8-1.png" style="display: block; margin: auto;" />

The `qqplotr` package also offers the option to *detrend* Q-Q and P-P plots in order to help reducing visual bias caused by the orthogonal distances from points to the reference lines (Thode, 2002). That bias may cause wrong conclusions to be drawn via visual inference of the plot. To do that we must set `detrend = TRUE`:

``` r
di <- "exp"
dp <- list(rate = 2)
de <- TRUE # enabling the detrend option

gg <- ggplot(data = airquality, mapping = aes(sample = Ozone)) +
    stat_qq_band(distribution = di, dparams = dp, detrend = de) +
    stat_qq_line(distribution = di, dparams = dp, detrend = de) +
    stat_qq_point(distribution = di, dparams = dp, detrend = de) +
    labs(x = "Theoretical Quantiles", y = "Sample Quantiles")
gg
```

<img src="man/figures/README-unnamed-chunk-9-1.png" style="display: block; margin: auto;" />

Note that the detrend option causes to plot to "rotate", that is, instead of being presented as a diagonal plot, we visualize the data in a horizontal manner.

The Q-Q plot functions are also compatible with many `ggplot2` operators, such as Facets (sub-plots). For instance, lets consider the *barley* dataset from the `lattice` package to illustrate how Facets behave when applied to the Q-Q plot functions:

``` r
# install.packages("lattice")
data("barley", package = "lattice")

gg <- ggplot(data = barley, mapping = aes(sample = yield, color = site, fill = site)) +
    stat_qq_band() +
    stat_qq_line() +
    stat_qq_point() +
    facet_wrap(~ site) +
    labs(x = "Theoretical Quantiles", y = "Sample Quantiles")
gg
```

<img src="man/figures/README-unnamed-chunk-10-1.png" style="display: block; margin: auto;" />

### P-P plot

Let's start by plotting the previously simulated Normal data versus the standard Normal distribution:

``` r
gg <- ggplot(data = smp, mapping = aes(sample = norm)) +
    stat_pp_band() +
    stat_pp_line() +
    stat_pp_point() +
    labs(x = "Probability Points", y = "Cumulative Probability")
gg
```

<img src="man/figures/README-unnamed-chunk-11-1.png" style="display: block; margin: auto;" />

Notice that the label names are different from those of the Q-Q plots. Here, the cumulative probability points (y-axis) are constructed by evaluating the theoretical CDF on sample quantiles.

As discussed before, in the case of P-P plots the distributional parameters **do** impact the results. For instance, say we want to evaluate the same standard Normal data with a shifted and rescaled Normal(2,2) distribution:

``` r
dp <- list(mean = 2, sd = 2) # shifted and rescaled Normal parameters

gg <- ggplot(data = smp, mapping = aes(sample = norm)) +
    stat_pp_band(dparams = dp) +
    stat_pp_line() +
    stat_pp_point(dparams = dp) +
    labs(x = "Probability Points", y = "Cumulative Probability")
gg
```

<img src="man/figures/README-unnamed-chunk-12-1.png" style="display: block; margin: auto;" />

As we already know, the plot shows that the chosen Normal distribution parameters are not appropriate for the input data.

Also notice that the `stat_pp_line` function lacks the `dparams` paramater. The reason is that `stat_pp_line` draws by default the identity line and, thus, it isn't dependent of the sample data and/or its distribution. However, if the user wishes to draw another line (different from the identity) he/she may do so by providing the intercept and slope values, respectively, as a vector of length two to the `ab` parameter:

``` r
gg <- ggplot(data = smp, mapping = aes(sample = norm)) +
    stat_pp_band() +
    stat_pp_line(ab = c(.2, .5)) + # intercept = 0.2, slope = 0.5
    stat_pp_point() +
    labs(x = "Probability Points", y = "Cumulative Probability")
gg
```

<img src="man/figures/README-unnamed-chunk-13-1.png" style="display: block; margin: auto;" />

We may also detrend the P-P plots in the same way as before. Let's evaluate again the *mean ozone levels* from the `airquality` dataset. We had previously seen that the Exponential distribution was more appropriate than the Normal distribution, so let's take that into account:

``` r
di <- "exp"
dp <- list(rate = .022) # value is based on some empirical tests
de <- TRUE

gg <- ggplot(data = airquality, mapping = aes(sample = Ozone)) +
    stat_pp_band(distribution = di, detrend = de, dparams = dp) +
    stat_pp_line(detrend = de) +
    stat_pp_point(distribution = di, detrend = de, dparams = dp) +
    labs(x = "Probability Points", y = "Cumulative Probability") +
    scale_y_continuous(limits = c(-.5, .5))
gg
```

<img src="man/figures/README-unnamed-chunk-14-1.png" style="display: block; margin: auto;" />

Based on empirical tests, we set the rate parameter to `rate = .022`. That value let the most P-P points inside the confidence bands. Even so, that group of outside the confidence bands (at the lower tail) indicate that a more appropriate distribution should be selected.

Shiny App
---------

In this package, we also included an interactive Shiny app with which you're able to explore this package functions and its parameters. To run the app, simply call:

``` r
runShinyExample()
```

References
----------

-   [Thode, H. (2002), Testing for Normality. CRC Press, 1st Ed.](https://www.crcpress.com/Testing-For-Normality/Thode/p/book/9780824796136)
-   [Aldor-Noiman, S. et al. (2013). The Power to See: A New Graphical Test of Normality. The American Statistician. 67:4.](http://www.tandfonline.com/doi/abs/10.1080/00031305.2013.847865)


NEWS.md
# qqplotr 0.0.2

* Added the `identity` argument to `*_qq_*` functions, which permits to draw the
identity line (instead of the usual Q-Q line).

* Changed `"bs"` to `"boot"` as one of the possible values of the `bandType`
argument, for obvious reasons. :)

* By default, distributional parameters are now automatically estimated with MLE
for `stat_qq_point`, `stat_qq_line`, `stat_pp_point`, and `stat_pp_band`.

* Created a Shiny app to showcase the package functions and its parameters.

* Included error handling for most of the main parameters from all functions.

* Added partial match for most of the character arguments.

* Fixed a bug caused by `stat_qq_line` default colors when facetting.

* Created the vignette `introduction.Rmd`.

* Removed most inheritances from Q-Q functions, as they were not actually
essential.

* Added detrend option for P-P functions.

* Changed the P-P functions labels in the examples from the documentation.

* Implemented the P-P band stat `stat_pp_band` via parametric bootstrap.

* Created a description for the package itself, i.e., now `?qqplotr` works.

* Implemented the P-P plot stats: `stat_pp_point` and `stat_pp_line`.

# qqplotr 0.0.1

* Initial release of the package.

* Implementation of three ggplot2 stats: `stat_qq_point`, `stat_qq_line` and
`stat_qq_band`.


R/stat_qq_band.R
#' Quantile-quantile confidence bands
#'
#' Draws quantile-quantile confidence bands, with an additional detrend option.
#'
#' @import ggplot2
#' @importFrom dplyr summarize
#' @importFrom dplyr group_by
#' @importFrom MASS fitdistr
#' @importFrom robustbase s_Qn
#' @importFrom robustbase Qn
#'
#' @include stat_qq_line.R
#'
#' @inheritParams ggplot2::layer
#' @inheritParams ggplot2::geom_ribbon
#'
#' @param distribution Character. Theoretical probability distribution function
#'   to use. Do not provide the full distribution function name (e.g.,
#'   \code{"dnorm"}). Instead, just provide its shortened name (e.g.,
#'   \code{"norm"}). If you wish to provide a custom distribution, you may do so
#'   by first creating the density, quantile, and random functions following the
#'   standard nomenclature from the \code{stats} package (i.e., for
#'   \code{"custom"}, create the \code{dcustom}, \code{pcustom},
#'   \code{qcustom}, and \code{rcustom} functions).
#' @param dparams List of additional parameters passed on to the previously
#'   chosen \code{distribution} function. If an empty list is provided (default)
#'   then the distributional parameters are estimated via MLE. MLE for custom
#'   distributions is currently not supported, so you must provide the
#'   appropriate \code{dparams} in that case.
#' @param detrend Logical. Should the plot objects be detrended? If \code{TRUE},
#'   the objects will be detrended according to the reference Q-Q line. This
#'   procedure was described by Thode (2002), and may help reducing visual bias
#'   caused by the orthogonal distances from Q-Q points to the reference line.
#' @param identity Logical. Should an identity line be used as the reference
#'   line used to construct the confidence bands? If \code{TRUE}, the identity
#'   line is used. If \code{FALSE} (default), the commonly-used Q-Q line that
#'   intercepts two data quantiles specified in \code{qprobs} is used. Please
#'   notice that the chosen reference line will also be used for the detrending
#'   procedure, if \code{detrend = TRUE}.
#' @param qtype Integer between 1 and 9. Type of the quantile algorithm to be
#'   used by the \code{\link[stats]{quantile}} function to construct the Q-Q
#'   line.
#' @param qprobs Numeric vector of length two. Represents the quantiles used by
#'   the \code{\link[stats]{quantile}} function to construct the Q-Q line.
#' @param bandType Character. Either \code{"normal"}, \code{"boot"} or
#'   \code{"ts"}. \code{"normal"} constructs simultaneous confidence bands based
#'   on Normal confidence intervals. \code{"boot"} creates pointwise confidence
#'   bands based on a parametric bootstrap; parameters are estimated with MLEs.
#'   Finally, \code{"ts"} constructs tail-sensitive confidence bands, as
#'   described by Aldor-Noiman et al. (2013) (also, see 'Note' for
#'   limitations).
#' @param B Integer. If \code{bandType = "boot"}, then \code{B} is the number of
#'   bootstrap replicates. If \code{bandType = "ts"}, then \code{B} is the
#'   number of simulated samples.
#' @param conf Numerical. Confidence level of the bands.
#' @param mu Numerical. Only used if \code{bandType = "ts"}. Center
#'   distributional parameter used to construct the simulated tail-sensitive
#'   confidence bands. If either \code{mu} or \code{sigma} are \code{NULL}, then
#'   those parameters are estimated using \code{\link[robustbase]{Qn}} and
#'   \code{\link[robustbase]{s_Qn}}, respectively.
#' @param sigma Numerical. Only used if \code{bandType = "ts"}. Scale
#'   distributional parameter used to construct the simulated tail-sensitive
#'   confidence bands. If either \code{mu} or \code{sigma} are \code{NULL}, then
#'   those parameters are estimated using robust estimates from the \pkg{stats}
#'   package.
#'
#' @note
#' \itemize{
#' \item{Tail-sensitive confidence bands are only implemented for Normal Q-Q
#' plots. As a future update, we intend to generalize to other distributions.}
#' \item{Bootstrap bands are constructed based on a MLE parametric bootstrap.
#' Hence, it is not possible to construct such bands if the sample and
#' theoretical distributions present mismatching supports.}
#' }
#'
#' @references
#' \itemize{
#' \item{\href{https://www.crcpress.com/Testing-For-Normality/Thode/p/book/9780824796136}{Thode,
#' H. (2002), Testing for Normality. CRC Press, 1st Ed.}}
#' \item{\href{http://www.tandfonline.com/doi/abs/10.1080/00031305.2013.847865}{Aldor-Noiman,
#' S. et al. (2013). The Power to See: A New Graphical Test of Normality. The
#' American Statistician. 67:4.}}
#' }
#'
#' @examples
#' # generate random Normal data
#' set.seed(0)
#' smp <- data.frame(norm = rnorm(100))
#'
#' # Normal Q-Q plot of Normal data
#' gg <- ggplot(data = smp, mapping = aes(sample = norm)) +
#'  stat_qq_band() +
#'  stat_qq_line() +
#'  stat_qq_point()
#' gg + labs(x = "Theoretical Quantiles", y = "Sample Quantiles")
#'
#' # Exponential Q-Q plot of mean ozone levels (airquality dataset)
#' di <- "exp"
#' dp <- list(rate = 1)
#' gg <- ggplot(data = airquality, mapping = aes(sample = Ozone)) +
#'  stat_qq_band(distribution = di, dparams = dp) +
#'  stat_qq_line(distribution = di, dparams = dp) +
#'  stat_qq_point(distribution = di, dparams = dp) +
#'  labs(x = "Theoretical Quantiles", y = "Sample Quantiles")
#' gg
#'
#' # Detrended Exponential Q-Q plot of mean ozone levels
#' di <- "exp"
#' dp <- list(rate = 1)
#' de <- TRUE
#' gg <- ggplot(data = airquality, mapping = aes(sample = Ozone)) +
#'  stat_qq_band(distribution = di, detrend = de) +
#'  stat_qq_line(distribution = di, detrend = de) +
#'  stat_qq_point(distribution = di, detrend = de) +
#'  labs(x = "Theoretical Quantiles", y = "Sample Quantiles")
#' gg
#'
#' # Normal Q-Q plot of Normal data with boostrap confidence bands
#' bt <- "boot"
#' gg <- ggplot(data = smp, mapping = aes(sample = norm)) +
#'  stat_qq_band(bandType = bt) +
#'  stat_qq_line() +
#'  stat_qq_point() +
#'  labs(x = "Theoretical Quantiles", y = "Sample Quantiles")
#' gg
#'
#' # Normal Q-Q plot of Normal data with tail-sensitive confidence bands
#' bt <- "ts"
#' gg <- ggplot(data = smp, mapping = aes(sample = norm)) +
#'  stat_qq_band(bandType = bt) +
#'  stat_qq_line() +
#'  stat_qq_point() +
#'  labs(x = "Theoretical Quantiles", y = "Sample Quantiles")
#' gg
#'
#' @export
stat_qq_band <- function(data = NULL,
												 mapping = NULL,
												 geom = "qq_band",
												 position = "identity",
												 show.legend = NA,
												 inherit.aes = TRUE,
												 na.rm = TRUE,
												 distribution = "norm",
												 dparams = list(),
												 detrend = FALSE,
												 identity = FALSE,
												 qtype = 7,
												 qprobs = c(.25, .75),
												 bandType = "normal",
												 B = 1000,
												 conf = .95,
												 mu = NULL,
												 sigma = NULL,
												 ...) {
	# error handling
	if (!(distribution %in% c(
		"beta",
		"cauchy",
		"chisq",
		"exp",
		"f",
		"gamma",
		"geom",
		"lnorm",
		"logis",
		"norm",
		"nbinom",
		"pois",
		"t",
		"weibull")) &
		length(dparams) == 0 &
		table(sapply(formals(eval(parse(text = paste0("q", distribution)))), typeof))["symbol"] > 1) {
		stop(
			"MLE is currently not supported for custom distributions.\n",
			"Please provide all the custom distribution parameters to 'dparams'.",
			call. = FALSE
		)
	}
	if (qtype < 1 | qtype > 9) {
		stop("Please provide a valid quantile type: ",
				 "'qtype' must be between 1 and 9.",
				 call. = FALSE)
	}
	if (conf < 0 | conf > 1) {
		stop("Please provide a valid confidence level for the bands: ",
				 "'conf' must be between 0 and 1.",
				 call. = FALSE)
	}
	if (B < 0) {
		stop("Please provide a positive value for B.",
				 call. = FALSE)
	}

	# vector with common discrete distributions
	discreteDist <- c("binom", "geom", "nbinom", "pois")

	if (distribution %in% discreteDist) geom <- "errorbar"

	ggplot2::layer(
		data = data,
		mapping = mapping,
		stat = StatQqBand,
		geom = geom,
		position = position,
		show.legend = show.legend,
		inherit.aes = inherit.aes,
		params = list(
			na.rm = na.rm,
			distribution = distribution,
			dparams = dparams,
			detrend = detrend,
			identity = identity,
			qtype = qtype,
			qprobs = qprobs,
			bandType = match.arg(bandType, c("normal", "boot", "ts")),
			B = round(B),
			conf = conf,
			mu = mu,
			sigma = sigma,
			discrete = distribution %in% discreteDist,
			...
		)
	)
}

#' StatQqBand
#'
#' @keywords internal
#' @usage NULL
#' @export
StatQqBand <- ggplot2::ggproto(
	`_class` = "StatQqBand",
	`_inherit` = StatQqLine,

	default_aes = ggplot2::aes(
		x = ..x..,
		ymin = ..lower..,
		ymax = ..upper..
	),

	required_aes = c("sample"),

	compute_group = {
		function(data,
						 self,
						 scales,
						 distribution,
						 dparams,
						 detrend,
						 identity,
						 qtype,
						 qprobs,
						 bandType,
						 B,
						 conf,
						 mu,
						 sigma,
						 discrete) {
			# distributional functions
			dFunc <- eval(parse(text = paste0("d", distribution)))
			qFunc <- eval(parse(text = paste0("q", distribution)))
			rFunc <- eval(parse(text = paste0("r", distribution)))

			smp <- sort(data$sample)
			n <- length(smp)
			quantiles <- ppoints(n)

			# automatically estimate parameters with MLE, only if no parameters are
			# provided with dparams and there are at least one distributional parameter
			# without a default value
			if(length(dparams) == 0) {
				# equivalence between base R and MASS::fitdistr distribution names
				corresp <- function(distName) {
					switch(
						distName,
						beta = "beta",
						cauchy = "cauchy",
						chisq = "chi-squared",
						exp = "exponential",
						f = "f",
						gamma = "gamma",
						geom = "geometric",
						lnorm = "log-normal",
						logis = "logistic",
						norm = "normal",
						nbinom = "negative binomial",
						pois = "poisson",
						t = dt,
						weibull = "weibull",
						NULL
					)
				}

				# initial value for some distributions
				initVal <- function(distName) {
					switch(
						distName,
						beta = list(shape1 = 1, shape2 = 1),
						chisq = list(df = 1),
						f = list(df1 = 1, df2 = 2),
						t = list(df = 1),
						NULL
					)
				}

				suppressWarnings({
					if(!is.null(corresp(distribution))) {
						if(is.null(initVal(distribution))) {
							dparams <- MASS::fitdistr(x = smp, densfun = corresp(distribution))$estimate
						} else {
							dparams <- MASS::fitdistr(x = smp, densfun = corresp(distribution), start = initVal(distribution))$estimate
						}
					}
				})
			}

			theoretical <- do.call(qFunc, c(list(p = quantiles), dparams))

			# inherit from StatQqLine
			xline <- self$super()$compute_group(data = data,
																					distribution = distribution,
																					dparams = dparams,
																					qtype = qtype,
																					qprobs = qprobs,
																					identity = identity,
																					detrend = FALSE)$xline
			yline <- self$super()$compute_group(data = data,
																					distribution = distribution,
																					dparams = dparams,
																					qtype = qtype,
																					qprobs = qprobs,
																					identity = identity,
																					detrend = FALSE)$yline

			slope <- diff(yline) / diff(xline)
			intercept <- yline[1L] - slope * xline[1L]

			fittedValues <- (slope * theoretical) + intercept

			# confidence bands based on normal confidence intervals
			if (bandType == "normal") {
				probs <- ppoints(n)
				stdErr <- (slope / do.call(dFunc, c(list(x = theoretical), dparams))) * sqrt(probs * (1 - probs) / n)
				zCrit <- qnorm(p = (1 - (1 - conf) / 2))

				upper <- fittedValues + (stdErr * zCrit)
				lower <- fittedValues - (stdErr * zCrit)
			}

			# parametric bootstrap pointwise confidence intervals
			if (bandType == "boot") {
				bs <- apply(
					X = matrix(do.call(rFunc, c(list(n = n * B), as.list(dparams))), n, B),
					MARGIN = 2,
					FUN = sort
				)

				upper <- apply(X = bs, MARGIN = 1, FUN = quantile, probs = (1 + conf) / 2)
				lower <- apply(X = bs, MARGIN = 1, FUN = quantile, probs = (1 - conf) / 2)
			}

			# tail-sensitive confidence bands
			if (bandType == "ts") {
				if (distribution != "norm") {
					warning("Be aware that tail-sensitive confidence bands are _only_ implemented for Normal Q-Q plots.",
									call. = F)
				}

				centerFunc <- function(x) robustbase::s_Qn(x, mu.too = TRUE)[[1]]
				scaleFunc <- function(x) robustbase::Qn(x, finite.corr = FALSE)

				upperCi <- rep(NA, n)
				lowerCi <- rep(NA, n)
				pValue <- matrix(NA, nrow = n, ncol = B)

				# simulate data
				sim <- NULL
				if (is.null(mu) | is.null(sigma)) {
					for (i in 1:B) sim <- cbind(sim, sort(rnorm(n)))

					# center and scale simulated data
					center <- apply(sim, 2, centerFunc)
					scale <- apply(sim, 2, scaleFunc)
					sim <- sweep(sweep(sim, 2, center, FUN = "-"), 2, scale, FUN = "/")

					# convert simulated values to probabilities
					sim <- t(apply(sim, 1, pnorm))
				} else {
					for (i in 1:B) sim <- cbind(sim, sort(runif(n)))
				}

				# widen the CIs to get simultanoues (100 * conf)% CIs
				for (i in 1:n) {
					tmp <- pbeta(sim[i, ], shape1 = i, shape2 = n + 1 - i)
					pValue[i, ] <- apply(cbind(tmp, 1 - tmp), 1, min)
				}

				critical <- apply(pValue, 2, min)
				criticalC <- quantile(critical, prob = 1 - conf)

				upperCi <- qbeta(1 - criticalC, shape1 = 1:n, shape2 = n + 1 - (1:n))
				lowerCi <- qbeta(criticalC, shape1 = 1:n, shape2 = n + 1 - (1:n))

				# translate back to sample quantiles
				if (is.null(mu) | is.null(sigma)) {
					upper <- qnorm(upperCi) * scaleFunc(smp) + centerFunc(smp)
					lower <- qnorm(lowerCi) * scaleFunc(smp) + centerFunc(smp)
				} else {
					upper <- qnorm(upperCi) * sigma + mu
					lower <- qnorm(lowerCi) * sigma + mu
				}
			}

			out <- data.frame(
				x = theoretical,
				upper = upper,
				lower = lower,
				fill = if (is.null(data$fill)) rgb(.6, .6, .6, .5) else data$fill
			)

			if (discrete) {
				out$colour <- rgb(.5, .5, .5)
				# create a data.frame with unique rows
				out <- dplyr::summarize(
								dplyr::group_by(out, x, fill, colour),
								upper = max(upper),
								lower = min(lower)
							 )
				out <- as.data.frame(out)
			}

			# detrend the confidence bands by keeping the same distance from the
			# stat_qq_line, which now is a line centered on y = 0
			if (detrend) {
				out$upper <- out$upper - fittedValues
				out$lower <- out$lower - fittedValues
			}

			out
		}
	}
)
